### PR TITLE
Support for guided decoding for offline LLM

### LINKED ISSUE #3536: [Feature]: Support Guided Decoding in `LLM` entrypoint
### 🚀 The feature, motivation and pitch

Currently we support guided decoding of (JSON, Regex, Choice, Grammar, and arbitrary JSON) in OpenAI inference server. It would be great that we expose the same functionality in the offline interface as well. 

https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#extra-parameters

Concretely this would mean adding the support here as a new parameter to generate call. Using methods introduced in #2819. 

Do make sure to add test and examples. 

### Alternatives

_No response_

### Additional context

_No response_

### LINKED ISSUE #7030: [Usage]: how do I pass in the JSON content-type for Mistral 7B using offline inference?
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

I would like to use the [JSON mode](https://docs.mistral.ai/capabilities/json_mode/) for Mistral 7B while doing offline inference using the `generate` method as below. Is that possible somehow? Just using the prompt doesn't seem to produce JSON output as requested. If this is not possible, is the only solution to use something like [Outlines](https://github.com/outlines-dev/outlines)? Would love some details on that, if so.

```
llm = LLM(model="mistralai/Mistral-7B-v0.3")
prompt = "Please name the biggest and smallest continent in JSON using the following schema: {biggest: <the biggest continent's name>, smallest: <the smallest continent>}"
sampling_params = SamplingParams(temperature=temperature, top_p=1.0)
response = self.llm.generate(prompt, sampling_params)
```

### PR DESCRIPTION
**Pull request by kevinbu233**
FILL IN THE PR DESCRIPTION HERE

Added support for guided decoding (JSON, regex, choice) on offline interface that is raised in issue https://github.com/vllm-project/vllm/issues/3536 . A refactor of https://github.com/vllm-project/vllm/pull/4130

FIX #3536
FIX #7030

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: tests/entrypoints/llm/conftest.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.fixture
+def sample_sql_statements():
+    return ("""
+start: select_statement
+select_statement: "SELECT" column "from" table "where" condition
+column: "col_1" | "col_2"
+table: "table_1" | "table_2"
+condition: column "=" number
+number: "1" | "2"
+""")
```
Comment (DarkLight1337) [inline tests/entrypoints/llm/conftest.py:69]: Perhaps we could share these fixtures between LLM and OpenAI tests by putting them in `tests/entrypoints/conftest.py`?

File: tests/entrypoints/llm/test_guided_generate.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    "The capital of France is",
+    "The future of AI is",
+]
+
+TOKEN_IDS = [
+    [0],
+    [0, 1],
+    [0, 2, 1],
+    [0, 3, 1, 2],
+]
```
Comment (DarkLight1337) [inline tests/entrypoints/llm/test_guided_generate.py:30]:
```suggestion

```

File: vllm/entrypoints/llm.py [hunk 3]
```diff
@@ -262,6 +265,7 @@ def generate(
         use_tqdm: bool = True,
         lora_request: Optional[Union[List[LoRARequest], LoRARequest]] = None,
         prompt_adapter_request: Optional[PromptAdapterRequest] = None,
+        guided_options: Optional[Union[Dict, GuidedDecodingFields]] = None
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:268]: We can use a `TypedDict` to better indicate what can be in the dictionary.

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -303,6 +307,13 @@ def generate(
         else:
             inputs = cast(Union[PromptInputs, Sequence[PromptInputs]], prompts)
 
+        if isinstance(guided_options, Dict):
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:310]:
```suggestion
        if isinstance(guided_options, dict):
```

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -303,6 +307,13 @@ def generate(
...
+            if len(guided_options) > 1:
+                raise ValueError(
+                    "You can only use one guided decoding but multiple is "
+                    f"specified: {self.__dict__}")
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:314]:
```suggestion
                raise ValueError(
                    "You can only use one guided decoding but multiple are "
                    f"specified: {guided_options}")
```

File: vllm/entrypoints/llm.py [hunk 6]
```diff
@@ -523,6 +536,18 @@ def _validate_and_add_requests(
             raise ValueError("The lengths of prompts and lora_request "
                              "must be the same.")
 
+        if isinstance(params, list):
+            if len(params) != num_requests:
+                raise ValueError("The lengths of prompts and params "
+                                 "must be the same.")
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:542]: Already checked above
```suggestion

```

File: vllm/entrypoints/llm.py [hunk 7]
```diff
@@ -548,6 +573,24 @@ def _add_request(
...
+                guided_options.guided_decoding_backend = (
+                    decoding_config.guided_decoding_backend)
+            guided_logits_processor = get_local_guided_decoding_logits_processor(  #noqa
+                guided_options.guided_decoding_backend, guided_options,
+                self.get_tokenizer())
+            if guided_logits_processor:
+                if params.logits_processors is None:
+                    params.logits_processors = []
+                params.logits_processors.append(guided_logits_processor)
+        return params
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:592]: Perhaps we should make this a method of `SamplingParams` instead of repeating the definition in LLM and OpenAI server?

File: vllm/model_executor/guided_decoding/guided_fields.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    guided_json_object: Optional[bool] = None
+
+    def __post_init__(self):
+        """Validate that some fields are mutually exclusive."""
+        guide_count = sum([
+            self.guided_json is not None,
+            self.guided_regex is not None,
+            self.guided_choice is not None,
+            self.guided_grammar is not None,
+            self.guided_json_object is not None, ])
```
Comment (DarkLight1337) [inline vllm/model_executor/guided_decoding/guided_fields.py:25]:
```suggestion
            self.guided_json_object is not None])
```

File: vllm/model_executor/guided_decoding/guided_fields.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        if guide_count > 1:
+            raise ValueError(
+                "You can only use one kind of guided decoding but multiple is "
```
Comment (DarkLight1337) [inline vllm/model_executor/guided_decoding/guided_fields.py:28]:
```suggestion
                "You can only use one kind of guided decoding but multiple are "
```

File: vllm/model_executor/guided_decoding/guided_fields.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+from pydantic import BaseModel
+
+
+@dataclass
+class GuidedDecodingFields:
```
Comment (DarkLight1337) [inline vllm/model_executor/guided_decoding/guided_fields.py:8]: Maybe rename this to `GuidedDecodingRequest` to be consistent with the other parameters to `LLM.generate`.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for implementing this! Left a few comments.

Also, it would be a good idea to implement this for `AsyncLLMEngine` so we can move the guided decoding logic out of the OpenAI server.

File: vllm/entrypoints/llm.py (hunk 3 above)
Comment (kevinbu233) [reply vllm/entrypoints/llm.py:268]: Thank you so much for the comments and the review! I really appreciate it! For this, I feel like the GuidedDecodingFields is already doing the job of specifying what are the available fields?

File: vllm/entrypoints/llm.py (hunk 7 above)
Comment (kevinbu233) [reply vllm/entrypoints/llm.py:592]: I currently want to separate out GuidedDecodingFields, which is used by local LLM, and ChatRequest and CompletetionRequest, used by openai server as they have different ways of retrieving the logits processors. As openai server's retriever of logits processor is async and local LLM is synchronous, it would probably still require separate methods even if it is implement in SamplingParams.
I have implemented all the other suggestions. I really appreciate the time you took to review them! If you have any further suggestions, do let me know!

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Alright, let's keep them as is for now.

File: vllm/entrypoints/llm.py (hunk 3 above)
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Using a `TypedDict` would enable auto-completion as opposed to just a dictionary with arbitrary keys/values.

Comment (kevinbu233) [reply vllm/entrypoints/llm.py]: That makes sense. Thank you!
I pushed a new commit with the TypedDict added. I wasn't sure where to put the definition. If you think it needs to be outside of llm.py, do let me know! Thank you!

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: I would put it next to the corresponding dataclass since this would make it easier to update both together.

Comment (kevinbu233) [reply vllm/entrypoints/llm.py]: Okok, fixed!

**Comment by roG0d** [comment]
Hi, I've been tracked the guided decoding with CFG for vLLM this past week. I found this PR that points exactly to what I'm looking for. 

To be on the same page: it's true that vLLM supports grammar but just with requests related to OpenAI Compatible Server?

If that's so, I would like to help and try bringing this feature forward.

**Review by DarkLight1337 — APPROVED** [review]
LGTM now. You just have to run `format.sh` to appease the linter.

**Comment by DarkLight1337** [comment]
> Hi, I've been tracked the guided decoding with CFG for vLLM this past week. I found this PR that points exactly to what I'm looking for.
> 
> To be on the same page: it's true that vLLM supports grammar but just with requests related to OpenAI Compatible Server?
> 
> If that's so, I would like to help and try bringing this feature forward.

Yes, this PR just extends the existing functionality to offline `LLM` class. The PR is basically complete already, thanks for offering help though!

**Comment by rkooo567** [comment]
QQ: should we deprecate logit processor API and pass the string to sampling parameter instead?

**Comment by DarkLight1337** [comment]
> QQ: should we deprecate logit processor API and pass the string to sampling parameter instead?

Could you elaborate on that? Not sure how would that work.

**Comment by stas00** [comment]
@kevinbu233, can this be merged? I think there are quite a few who have been waiting for this feature since many moons. Thank you!

**Comment by stas00** [comment]
Thank you for the quick merge!

**Comment by han-steve** [comment]
Can we add a docstring for the guided_options_request param [here](https://github.com/vllm-project/vllm/blob/e42c634acbd1b86b5becca51e8b8108a32a438d5/vllm/entrypoints/llm.py#L287)? It's pretty hard to discover this option on the documentation.
